File: src-tauri/src/printer.rs
use encoding_rs::GBK;
use serde::Deserialize;
use std::io::Write;
use std::net::{SocketAddr, TcpStream};
use std::time::Duration;

#[derive(Debug, Deserialize)]
pub struct PrinterConfig {
    pub mode: String,        // "serial" | "network"
    pub serial_port: String,
    pub baud_rate: u32,
    pub net_host: String,
    pub net_port: u16,
    pub encoding: String,   // "gbk" | "utf8"
}

#[derive(Debug, Deserialize)]
pub struct PrintOrderData {
    pub product: String,
    pub seq: i64,
    pub user_name: String,
    pub content: String,
    pub match_str: String,
    pub matched_at: i64, // unix ms
}

#[derive(Debug, Deserialize)]
pub struct TemplateField {
    pub id: String,
    pub visible: bool,
    pub align: String, // "left" | "center" | "right"
    pub bold: bool,
}

#[derive(Debug, Deserialize)]
pub struct PrintTemplate {
    pub title: String,
    pub paper_width: u8,
    pub fields: Vec<TemplateField>,
}

// ── Transport ────────────────────────────────────────────────

fn send_serial(port: &str, baud: u32, data: &[u8]) -> Result<(), String> {
    let mut port = serialport::new(port, baud)
        .timeout(Duration::from_secs(5))
        .open()
        .map_err(|e| format!("打开串口失败: {e}"))?;
    port.write_all(data).map_err(|e| format!("写入串口失败: {e}"))
}

fn send_network(host: &str, port: u16, data: &[u8]) -> Result<(), String> {
    let addr: SocketAddr = format!("{host}:{port}")
        .parse()
        .map_err(|_| format!("无效地址: {host}:{port}"))?;
    let mut stream = TcpStream::connect_timeout(&addr, Duration::from_secs(5))
        .map_err(|e| format!("连接打印机失败: {e}"))?;
    stream
        .write_all(data)
        .map_err(|e| format!("发送数据失败: {e}"))
}

fn dispatch(config: &PrinterConfig, data: &[u8]) -> Result<(), String> {
    match config.mode.as_str() {
        "network" => send_network(&config.net_host, config.net_port, data),
        _ => send_serial(&config.serial_port, config.baud_rate, data),
    }
}

// ── Encoding ─────────────────────────────────────────────────

fn encode_text(text: &str, encoding: &str) -> Vec<u8> {
    if encoding == "gbk" {
        let (encoded, _, _) = GBK.encode(text);
        encoded.into_owned()
    } else {
        text.as_bytes().to_vec()
    }
}

// ── Time helper ──────────────────────────────────────────────

fn format_unix_ms(ms: i64) -> String {
    let total_secs = ms / 1000 + 8 * 3600;
    let h = (total_secs / 3600) % 24;
    let m = (total_secs % 3600) / 60;
    let s = total_secs % 60;
    format!("{h:02}:{m:02}:{s:02}")
}

// ── ESC/POS receipt builder ──────────────────────────────────

fn build_receipt(lines: &[(String, String, bool)], encoding: &str) -> Vec<u8> {
    let mut buf: Vec<u8> = Vec::new();

    // ESC @ — initialize
    buf.extend_from_slice(&[0x1B, 0x40]);

    for (align, text, bold) in lines {
        let a: u8 = match align.as_str() {
            "center" => 1,
            "right" => 2,
            _ => 0,
        };
        // ESC a n — set justification
        buf.extend_from_slice(&[0x1B, 0x61, a]);
        if *bold {
            buf.extend_from_slice(&[0x1B, 0x45, 1]); // ESC E 1 — bold on
        }
        buf.extend(encode_text(text, encoding));
        buf.push(b'\n');
        if *bold {
            buf.extend_from_slice(&[0x1B, 0x45, 0]); // ESC E 0 — bold off
        }
    }

    // Feed 3 lines + partial cut
    buf.extend_from_slice(&[b'\n', b'\n', b'\n']);
    buf.extend_from_slice(&[0x1D, 0x56, 0x42, 3]); // GS V 66 3 — partial cut

    buf
}

fn receipt_lines(template: &PrintTemplate, order: Option<&PrintOrderData>) -> Vec<(String, String, bool)> {
    let col = if template.paper_width == 58 { 32usize } else { 42usize };
    let sep = "-".repeat(col);
    let mut lines: Vec<(String, String, bool)> = Vec::new();

    lines.push(("center".to_string(), template.title.clone(), false));
    lines.push(("left".to_string(), sep.clone(), false));

    for field in &template.fields {
        if !field.visible {
            continue;
        }

        if field.id == "separator" {
            lines.push((field.align.clone(), sep.clone(), field.bold));
            continue;
        }

        let text = match field.id.as_str() {
            "product" => match order {
                Some(o) => format!("商品: {}", o.product),
                None => "商品: 测试商品".to_string(),
            },
            "user_name" => match order {
                Some(o) => format!("用户: {}", o.user_name),
                None => "用户: 测试用户".to_string(),
            },
            "content" => match order {
                Some(o) => format!("弹幕: {}", o.content),
                None => "弹幕: 弹幕内容测试".to_string(),
            },
            "match_str" => match order {
                Some(o) => format!("匹配: {}  #{}", o.match_str, o.seq),
                None => "匹配: 测试关键词  #1".to_string(),
            },
            "seq" => match order {
                Some(o) => format!("序号: #{}", o.seq),
                None => "序号: #1".to_string(),
            },
            "time" => match order {
                Some(o) => format!("时间: {}", format_unix_ms(o.matched_at)),
                None => "时间: 12:00:00".to_string(),
            },
            _ => continue,
        };

        lines.push((field.align.clone(), text, field.bold));
    }

    lines
}

// ── Tauri commands ───────────────────────────────────────────

#[tauri::command]
pub fn list_serial_ports() -> Vec<String> {
    serialport::available_ports()
        .unwrap_or_default()
        .into_iter()
        .map(|p| p.port_name)
        .collect()
}

#[tauri::command]
pub fn printer_test(config: PrinterConfig, template: PrintTemplate) -> Result<(), String> {
    let lines = receipt_lines(&template, None);
    let data = build_receipt(&lines, &config.encoding);
    dispatch(&config, &data)
}

#[tauri::command]
pub fn printer_print_order(
    config: PrinterConfig,
    order: PrintOrderData,
    template: PrintTemplate,
) -> Result<(), String> {
    let lines = receipt_lines(&template, Some(&order));
    let data = build_receipt(&lines, &config.encoding);
    dispatch(&config, &data)
}

File: electron/main/modules/printer.ts
import { SerialPort } from 'serialport'
import * as net from 'node:net'
import iconvLite from 'iconv-lite'

export interface PrinterConfig {
  mode: 'serial' | 'network'
  serial_port: string
  baud_rate: number
  net_host: string
  net_port: number
  encoding: 'gbk' | 'utf8'
}

export interface PrintOrderData {
  product: string
  seq: number
  user_name: string
  content: string
  match_str: string
  matched_at: number
}

export interface TemplateField {
  id: string
  visible: boolean
  align: 'left' | 'center' | 'right'
  bold: boolean
}

export interface PrintTemplate {
  title: string
  paper_width: 58 | 80
  fields: TemplateField[]
}

// ── Transport ────────────────────────────────────────────────

async function sendSerial(portName: string, baudRate: number, data: Buffer): Promise<void> {
  return new Promise((resolve, reject) => {
    const port = new SerialPort({ path: portName, baudRate, autoOpen: false })
    port.open(err => {
      if (err) {
        reject(new Error(`打开串口失败: ${err.message}`))
        return
      }
      port.write(data, writeErr => {
        if (writeErr) {
          port.close()
          reject(new Error(`写入串口失败: ${writeErr.message}`))
          return
        }
        port.drain(drainErr => {
          port.close()
          if (drainErr) {
            reject(new Error(`串口drain失败: ${drainErr.message}`))
          } else {
            resolve()
          }
        })
      })
    })
  })
}

async function sendNetwork(host: string, port: number, data: Buffer): Promise<void> {
  return new Promise((resolve, reject) => {
    const socket = net.createConnection({ host, port }, () => {
      socket.write(data, err => {
        if (err) {
          socket.destroy()
          reject(new Error(`发送数据失败: ${err.message}`))
          return
        }
        socket.end()
        resolve()
      })
    })
    socket.setTimeout(5000, () => {
      socket.destroy(new Error('连接超时'))
    })
    socket.on('error', (err: Error) => reject(new Error(`连接打印机失败: ${err.message}`)))
  })
}

async function dispatch(config: PrinterConfig, data: Buffer): Promise<void> {
  if (config.mode === 'network') {
    return sendNetwork(config.net_host, config.net_port, data)
  }
  return sendSerial(config.serial_port, config.baud_rate, data)
}

// ── Encoding ─────────────────────────────────────────────────

function encodeText(text: string, encoding: string): Buffer {
  if (encoding === 'gbk') {
    return iconvLite.encode(text, 'GBK')
  }
  return Buffer.from(text, 'utf-8')
}

// ── Time helper ──────────────────────────────────────────────

function formatUnixMs(ms: number): string {
  const totalSecs = Math.floor(ms / 1000) + 8 * 3600
  const h = Math.floor((totalSecs / 3600) % 24)
  const m = Math.floor((totalSecs % 3600) / 60)
  const s = Math.floor(totalSecs % 60)
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`
}

// ── ESC/POS receipt builder ──────────────────────────────────

function buildReceipt(lines: Array<[string, string, boolean]>, encoding: string): Buffer {
  const chunks: Buffer[] = []

  // ESC @ — initialize
  chunks.push(Buffer.from([0x1b, 0x40]))

  for (const [align, text, bold] of lines) {
    const a = align === 'center' ? 1 : align === 'right' ? 2 : 0
    // ESC a n — set justification
    chunks.push(Buffer.from([0x1b, 0x61, a]))
    if (bold) {
      chunks.push(Buffer.from([0x1b, 0x45, 1])) // ESC E 1 — bold on
    }
    chunks.push(encodeText(text, encoding))
    chunks.push(Buffer.from([0x0a])) // \n
    if (bold) {
      chunks.push(Buffer.from([0x1b, 0x45, 0])) // ESC E 0 — bold off
    }
  }

  // Feed 3 lines + partial cut
  chunks.push(Buffer.from([0x0a, 0x0a, 0x0a]))
  chunks.push(Buffer.from([0x1d, 0x56, 0x42, 3])) // GS V 66 3 — partial cut

  return Buffer.concat(chunks)
}

function receiptLines(
  template: PrintTemplate,
  order: PrintOrderData | null
): Array<[string, string, boolean]> {
  const col = template.paper_width === 58 ? 32 : 42
  const sep = '-'.repeat(col)
  const lines: Array<[string, string, boolean]> = []

  lines.push(['center', template.title, false])
  lines.push(['left', sep, false])

  for (const field of template.fields) {
    if (!field.visible) continue

    if (field.id === 'separator') {
      lines.push([field.align, sep, field.bold])
      continue
    }

    let text: string
    switch (field.id) {
      case 'product':
        text = order ? `商品: ${order.product}` : '商品: 测试商品'
        break
      case 'user_name':
        text = order ? `用户: ${order.user_name}` : '用户: 测试用户'
        break
      case 'content':
        text = order ? `弹幕: ${order.content}` : '弹幕: 弹幕内容测试'
        break
      case 'match_str':
        text = order ? `匹配: ${order.match_str}  #${order.seq}` : '匹配: 测试关键词  #1'
        break
      case 'seq':
        text = order ? `序号: #${order.seq}` : '序号: #1'
        break
      case 'time':
        text = order ? `时间: ${formatUnixMs(order.matched_at)}` : '时间: 12:00:00'
        break
      default:
        continue
    }

    lines.push([field.align, text, field.bold])
  }

  return lines
}

// ── Commands ─────────────────────────────────────────────────

export async function listSerialPorts(): Promise<string[]> {
  const ports = await SerialPort.list()
  return ports.map(p => p.path)
}

export async function printerTest(config: PrinterConfig, template: PrintTemplate): Promise<void> {
  const lines = receiptLines(template, null)
  const data = buildReceipt(lines, config.encoding)
  await dispatch(config, data)
}

export async function printerPrintOrder(
  config: PrinterConfig,
  order: PrintOrderData,
  template: PrintTemplate
): Promise<void> {
  const lines = receiptLines(template, order)
  const data = buildReceipt(lines, config.encoding)
  await dispatch(config, data)
}
